# MSDS692 Project: Predicting the Number of Car Accidents in Denver
## Natalia Weakly 
### Daily Accident data
### Artificial Neural Network

In [1]:
# Imports
import pandas as pd
import numpy as np
import datetime
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#Load csv file
#data previousely processed with MSDS692_IndividualAccidents.ipynb and saved as .csv file
accidents_ann=pd.read_csv('../MSDSProject/accidents_all_withIDs.csv',index_col=0, parse_dates=[1, 4])

In [3]:
#Check load results
accidents_ann.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 120056 entries, 0 to 120055
Columns: 128 entries, OBJECTID_1 to DayOfWeek_6
dtypes: float64(13), int64(111), object(4)
memory usage: 118.2+ MB


In [4]:
accidents_ann.head()

OBJECTID_1 FIRST_OCCURENCE_DATE     GEO_LON            GEO_LAT  \
0      41743  2014-01-01 00:40:00 -104.993603         39.7450192   
1      41745  2014-01-01 01:35:00 -104.969690  39.74002460000001   
2      41746  2014-01-01 02:04:00 -104.988533         39.7405637   
3      41748  2014-01-01 03:30:00 -105.034616         39.7585195   
4      41749  2014-01-01 03:50:00 -104.809755         39.7903534   

                  Date  BicycleInvolved  PedestrianInvolved  Holiday  \
0  2014-01-01 00:00:00                0                   0        1   
1  2014-01-01 00:00:00                0                   0        1   
2  2014-01-01 00:00:00                0                   0        1   
3  2014-01-01 00:00:00                0                   0        1   
4  2014-01-01 00:00:00                0                   0        1   

   RushHour  Daytime     ...       Neighboorhood_Westwood  \
0         0        0     ...                            0   
1         0        0     ...                            0   
2         0        0     ...                            0   
3         0        0     ...                            0   
4         0        0     ...                            0   

   Neighboorhood_Whittier  Neighboorhood_Windsor  DayOfWeek_0  DayOfWeek_1  \
0                       0                      0            0            0   
1                       0                      0            0            0   
2                       0                      0            0            0   
3                       0                      0            0            0   
4                       0                      0            0            0   

   DayOfWeek_2  DayOfWeek_3  DayOfWeek_4  DayOfWeek_5  DayOfWeek_6  
0            1            0            0            0            0  
1            1            0            0            0            0  
2            1            0            0            0            0  
3            1            0            0            0            0  
4            1            0            0            0            0  

[5 rows x 128 columns]

In [5]:
accidents_ann.tail()

OBJECTID_1 FIRST_OCCURENCE_DATE     GEO_LON     GEO_LAT  \
120051     166816  2018-12-31 11:50:00 -104.963127   39.739432   
120052     166817  2018-12-31 23:28:00 -104.960163  39.7271141   
120053     166929  2018-12-22 20:15:00 -104.977531  39.7386568   
120054     167001  2018-12-30 02:09:00 -104.903460  39.7584083   
120055     167000  2018-12-24 08:00:00 -104.898687  39.6522116   

                       Date  BicycleInvolved  PedestrianInvolved  Holiday  \
120051  2018-12-31 00:00:00                0                   0        0   
120052  2018-12-31 00:00:00                0                   0        0   
120053  2018-12-22 00:00:00                0                   0        0   
120054  2018-12-30 00:00:00                0                   0        0   
120055  2018-12-24 00:00:00                0                   0        0   

        RushHour  Daytime     ...       Neighboorhood_Westwood  \
120051         0        1     ...                            0   
120052         0        0     ...                            0   
120053         0        0     ...                            0   
120054         0        0     ...                            0   
120055         1        1     ...                            0   

        Neighboorhood_Whittier  Neighboorhood_Windsor  DayOfWeek_0  \
120051                       0                      0            1   
120052                       0                      0            1   
120053                       0                      0            0   
120054                       0                      0            0   
120055                       0                      0            1   

        DayOfWeek_1  DayOfWeek_2  DayOfWeek_3  DayOfWeek_4  DayOfWeek_5  \
120051            0            0            0            0            0   
120052            0            0            0            0            0   
120053            0            0            0            0            1   
120054            0            0            0            0            0   
120055            0            0            0            0            0   

        DayOfWeek_6  
120051            0  
120052            0  
120053            0  
120054            1  
120055            0  

[5 rows x 128 columns]

In [6]:
#copy columns into a new dataframe 
accidents_ann2 = accidents_ann[['Date', 'Holiday', 'DIA_max_temp', 'DIA_min_temp', \
                                'DIA_precipitation', 'DIA_snowfall','St_max_temp', 'St_min_temp',  'St_precipitation', \
                                'St_snowfall', 'DW_max_temp', 'DW_min_temp', 'DW_precipitation', 'DW_snowfall', 'DayOfWeek_0', \
                                'DayOfWeek_1', 'DayOfWeek_2', 'DayOfWeek_3', 'DayOfWeek_4', 'DayOfWeek_5', 'DayOfWeek_6']]

In [7]:
# Check results
accidents_ann2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 120056 entries, 0 to 120055
Data columns (total 21 columns):
Date                 120056 non-null object
Holiday              120056 non-null int64
DIA_max_temp         120056 non-null float64
DIA_min_temp         120056 non-null float64
DIA_precipitation    120056 non-null float64
DIA_snowfall         120056 non-null float64
St_max_temp          120056 non-null float64
St_min_temp          120056 non-null float64
St_precipitation     120056 non-null float64
St_snowfall          120056 non-null float64
DW_max_temp          120056 non-null float64
DW_min_temp          120056 non-null float64
DW_precipitation     120056 non-null float64
DW_snowfall          120056 non-null float64
DayOfWeek_0          120056 non-null int64
DayOfWeek_1          120056 non-null int64
DayOfWeek_2          120056 non-null int64
DayOfWeek_3          120056 non-null int64
DayOfWeek_4          120056 non-null int64
DayOfWeek_5          120056 non-null int64
DayO

In [8]:
accidents_ann2.head()

Date  Holiday  DIA_max_temp  DIA_min_temp  \
0  2014-01-01 00:00:00        1      0.396226      0.359551   
1  2014-01-01 00:00:00        1      0.396226      0.359551   
2  2014-01-01 00:00:00        1      0.396226      0.359551   
3  2014-01-01 00:00:00        1      0.396226      0.359551   
4  2014-01-01 00:00:00        1      0.396226      0.359551   

   DIA_precipitation  DIA_snowfall  St_max_temp  St_min_temp  \
0                0.1           0.0     0.548077      0.52381   
1                0.1           0.0     0.548077      0.52381   
2                0.1           0.0     0.548077      0.52381   
3                0.1           0.0     0.548077      0.52381   
4                0.1           0.0     0.548077      0.52381   

   St_precipitation  St_snowfall     ...       DW_min_temp  DW_precipitation  \
0               0.0          0.0     ...          0.425287               0.0   
1               0.0          0.0     ...          0.425287               0.0   
2               0.0          0.0     ...          0.425287               0.0   
3               0.0          0.0     ...          0.425287               0.0   
4               0.0          0.0     ...          0.425287               0.0   

   DW_snowfall  DayOfWeek_0  DayOfWeek_1  DayOfWeek_2  DayOfWeek_3  \
0          0.0            0            0            1            0   
1          0.0            0            0            1            0   
2          0.0            0            0            1            0   
3          0.0            0            0            1            0   
4          0.0            0            0            1            0   

   DayOfWeek_4  DayOfWeek_5  DayOfWeek_6  
0            0            0            0  
1            0            0            0  
2            0            0            0  
3            0            0            0  
4            0            0            0  

[5 rows x 21 columns]

In [9]:
# Calculate the daily number of accidents
accidents_anndaily = accidents_ann2.groupby('Date')[['Date']].count()

In [10]:
# check results
accidents_anndaily.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1826 entries, 2014-01-01 00:00:00 to 2018-12-31 00:00:00
Data columns (total 1 columns):
Date    1826 non-null int64
dtypes: int64(1)
memory usage: 28.5+ KB


In [11]:
accidents_anndaily.head()

Date
Date                     
2014-01-01 00:00:00    45
2014-01-02 00:00:00    70
2014-01-03 00:00:00    59
2014-01-04 00:00:00    72
2014-01-05 00:00:00    72

In [12]:
accidents_anndaily.tail()

Date
Date                     
2018-12-27 00:00:00    52
2018-12-28 00:00:00    46
2018-12-29 00:00:00    45
2018-12-30 00:00:00    46
2018-12-31 00:00:00    73

In [13]:
# df with records corresponding to each individual accident
# => duplicate records for the same day
accidents_ann2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 120056 entries, 0 to 120055
Data columns (total 21 columns):
Date                 120056 non-null object
Holiday              120056 non-null int64
DIA_max_temp         120056 non-null float64
DIA_min_temp         120056 non-null float64
DIA_precipitation    120056 non-null float64
DIA_snowfall         120056 non-null float64
St_max_temp          120056 non-null float64
St_min_temp          120056 non-null float64
St_precipitation     120056 non-null float64
St_snowfall          120056 non-null float64
DW_max_temp          120056 non-null float64
DW_min_temp          120056 non-null float64
DW_precipitation     120056 non-null float64
DW_snowfall          120056 non-null float64
DayOfWeek_0          120056 non-null int64
DayOfWeek_1          120056 non-null int64
DayOfWeek_2          120056 non-null int64
DayOfWeek_3          120056 non-null int64
DayOfWeek_4          120056 non-null int64
DayOfWeek_5          120056 non-null int64
DayO

In [14]:
#Drop duplicate records for the same-day observations and reset the index
accidents_ann2.drop_duplicates(inplace=True)
accidents_ann2.reset_index(inplace=True, drop=True)

C:\Users\RodneyWeakly\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [15]:
#check results
accidents_ann2.head()

Date  Holiday  DIA_max_temp  DIA_min_temp  \
0  2014-01-01 00:00:00        1      0.396226      0.359551   
1  2014-01-02 00:00:00        0      0.405660      0.337079   
2  2014-01-27 00:00:00        0      0.207547      0.314607   
3  2014-01-28 00:00:00        0      0.320755      0.202247   
4  2014-01-03 00:00:00        0      0.594340      0.494382   

   DIA_precipitation  DIA_snowfall  St_max_temp  St_min_temp  \
0           0.100000      0.000000     0.548077     0.523810   
1           0.000000      0.000000     0.355769     0.404762   
2           0.050000      0.022901     0.557692     0.369048   
3           0.000056      0.000000     0.201923     0.238095   
4           0.000000      0.000000     0.442308     0.369048   

   St_precipitation  St_snowfall     ...       DW_min_temp  DW_precipitation  \
0          0.000000     0.000000     ...          0.425287          0.000000   
1          0.016129     0.032520     ...          0.344828          0.033708   
2          0.000040     0.000008     ...          0.344828          0.000056   
3          0.020161     0.065041     ...          0.229885          0.061798   
4          0.000000     0.000000     ...          0.356322          0.000000   

   DW_snowfall  DayOfWeek_0  DayOfWeek_1  DayOfWeek_2  DayOfWeek_3  \
0     0.000000            0            0            1            0   
1     0.035294            0            0            0            1   
2     0.000012            1            0            0            0   
3     0.152941            0            1            0            0   
4     0.000000            0            0            0            0   

   DayOfWeek_4  DayOfWeek_5  DayOfWeek_6  
0            0            0            0  
1            0            0            0  
2            0            0            0  
3            0            0            0  
4            1            0            0  

[5 rows x 21 columns]

In [16]:
#merge observations for each day 
accidents_anndaily3 = pd.merge(accidents_anndaily, accidents_ann2, how='left', left_on=accidents_anndaily.index, right_on='Date') 

In [17]:
#check results
accidents_anndaily3.head()

Date  Date_x               Date_y  Holiday  DIA_max_temp  \
0  2014-01-01 00:00:00      45  2014-01-01 00:00:00        1      0.396226   
1  2014-01-02 00:00:00      70  2014-01-02 00:00:00        0      0.405660   
2  2014-01-03 00:00:00      59  2014-01-03 00:00:00        0      0.594340   
3  2014-01-04 00:00:00      72  2014-01-04 00:00:00        0      0.367925   
4  2014-01-05 00:00:00      72  2014-01-05 00:00:00        0      0.122642   

   DIA_min_temp  DIA_precipitation  DIA_snowfall  St_max_temp  St_min_temp  \
0      0.359551           0.100000           0.0     0.548077     0.523810   
1      0.337079           0.000000           0.0     0.355769     0.404762   
2      0.494382           0.000000           0.0     0.442308     0.369048   
3      0.280899           0.116667           0.0     0.605769     0.440476   
4      0.101124           0.027778           0.0     0.211538     0.250000   

      ...       DW_min_temp  DW_precipitation  DW_snowfall  DayOfWeek_0  \
0     ...          0.425287          0.000000     0.000000            0   
1     ...          0.344828          0.033708     0.035294            0   
2     ...          0.356322          0.000000     0.000000            0   
3     ...          0.482759          0.044944     0.141176            0   
4     ...          0.287356          0.056180     0.258824            0   

   DayOfWeek_1  DayOfWeek_2  DayOfWeek_3  DayOfWeek_4  DayOfWeek_5  \
0            0            1            0            0            0   
1            0            0            1            0            0   
2            0            0            0            1            0   
3            0            0            0            0            1   
4            0            0            0            0            0   

   DayOfWeek_6  
0            0  
1            0  
2            0  
3            0  
4            1  

[5 rows x 23 columns]

In [18]:
accidents_anndaily3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1826 entries, 0 to 1825
Data columns (total 23 columns):
Date                 1826 non-null object
Date_x               1826 non-null int64
Date_y               1826 non-null object
Holiday              1826 non-null int64
DIA_max_temp         1826 non-null float64
DIA_min_temp         1826 non-null float64
DIA_precipitation    1826 non-null float64
DIA_snowfall         1826 non-null float64
St_max_temp          1826 non-null float64
St_min_temp          1826 non-null float64
St_precipitation     1826 non-null float64
St_snowfall          1826 non-null float64
DW_max_temp          1826 non-null float64
DW_min_temp          1826 non-null float64
DW_precipitation     1826 non-null float64
DW_snowfall          1826 non-null float64
DayOfWeek_0          1826 non-null int64
DayOfWeek_1          1826 non-null int64
DayOfWeek_2          1826 non-null int64
DayOfWeek_3          1826 non-null int64
DayOfWeek_4          1826 non-null int64
DayOfW

In [19]:
#rename a column
accidents_anndaily3.rename(columns={'Date_x':'NumberOfAccidents'}, inplace=True)
# Drop 'Date_y' column
accidents_anndaily3.drop('Date_y', axis=1, inplace=True)
#Drop 'Date' column
accidents_anndaily3.drop('Date', axis=1, inplace=True)
#check results
accidents_anndaily3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1826 entries, 0 to 1825
Data columns (total 21 columns):
NumberOfAccidents    1826 non-null int64
Holiday              1826 non-null int64
DIA_max_temp         1826 non-null float64
DIA_min_temp         1826 non-null float64
DIA_precipitation    1826 non-null float64
DIA_snowfall         1826 non-null float64
St_max_temp          1826 non-null float64
St_min_temp          1826 non-null float64
St_precipitation     1826 non-null float64
St_snowfall          1826 non-null float64
DW_max_temp          1826 non-null float64
DW_min_temp          1826 non-null float64
DW_precipitation     1826 non-null float64
DW_snowfall          1826 non-null float64
DayOfWeek_0          1826 non-null int64
DayOfWeek_1          1826 non-null int64
DayOfWeek_2          1826 non-null int64
DayOfWeek_3          1826 non-null int64
DayOfWeek_4          1826 non-null int64
DayOfWeek_5          1826 non-null int64
DayOfWeek_6          1826 non-null int64
dtypes: 

In [20]:
accidents_anndaily3.head()

NumberOfAccidents  Holiday  DIA_max_temp  DIA_min_temp  DIA_precipitation  \
0                 45        1      0.396226      0.359551           0.100000   
1                 70        0      0.405660      0.337079           0.000000   
2                 59        0      0.594340      0.494382           0.000000   
3                 72        0      0.367925      0.280899           0.116667   
4                 72        0      0.122642      0.101124           0.027778   

   DIA_snowfall  St_max_temp  St_min_temp  St_precipitation  St_snowfall  \
0           0.0     0.548077     0.523810          0.000000     0.000000   
1           0.0     0.355769     0.404762          0.016129     0.032520   
2           0.0     0.442308     0.369048          0.000000     0.000000   
3           0.0     0.605769     0.440476          0.000040     0.000008   
4           0.0     0.211538     0.250000          0.060484     0.178862   

      ...       DW_min_temp  DW_precipitation  DW_snowfall  DayOfWeek_0  \
0     ...          0.425287          0.000000     0.000000            0   
1     ...          0.344828          0.033708     0.035294            0   
2     ...          0.356322          0.000000     0.000000            0   
3     ...          0.482759          0.044944     0.141176            0   
4     ...          0.287356          0.056180     0.258824            0   

   DayOfWeek_1  DayOfWeek_2  DayOfWeek_3  DayOfWeek_4  DayOfWeek_5  \
0            0            1            0            0            0   
1            0            0            1            0            0   
2            0            0            0            1            0   
3            0            0            0            0            1   
4            0            0            0            0            0   

   DayOfWeek_6  
0            0  
1            0  
2            0  
3            0  
4            1  

[5 rows x 21 columns]

## Use ANN (Artificial Neural Network) to forecast the daily number of accidents

In [25]:
#additional imports
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPRegressor

In [22]:
#set up target variable and independent variables
X=accidents_anndaily3.drop('NumberOfAccidents', axis=1) #independent variables
y=accidents_anndaily3['NumberOfAccidents'] #target variable

In [23]:
#Split the dataset into training and test part
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [26]:
#Create an instance of the model
mlp=MLPRegressor(hidden_layer_sizes=(20,20,20,20), max_iter=500)

In [27]:
# fit training data
mlp.fit(X_train, y_train)

MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(20, 20, 20, 20), learning_rate='constant',
       learning_rate_init=0.001, max_iter=500, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=None, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [28]:
#generate predictions
predictions=mlp.predict(X_test)

In [29]:
#Evaluate prediction results
#R2 score
mlp.score(X_test, y_test) 

0.3770237945738536

In [30]:
#Evaluate results
import sklearn.metrics

In [31]:
#Explained variance
sklearn.metrics.explained_variance_score(y_test, predictions)

0.3771447512842323

In [32]:
# Mean absolute error
sklearn.metrics.mean_absolute_error(y_test, predictions)

9.854272298967958

In [33]:
# Mean squared error
sklearn.metrics.mean_squared_error(y_test, predictions)

167.60375926368715

In [34]:
# Median absolute error
sklearn.metrics.median_absolute_error(y_test, predictions)

7.744360407248724

In [35]:
#R2, coefficient of determination, describes how well future observations are likely to be predicted [-1, 1]
sklearn.metrics.r2_score(y_test, predictions) 

0.3770237945738536


Overall, the model showed a relatively low predictive power, as according to the coefficient of determination it is likely to predict 37.7% of the future variations of the dependent variable.
Adjusting the model parameters (number of hidden layers, neurons, learning rate, the maximum number of iterations) did have any significant influence on the model's predictive power.
